rooms.js

```diff
@@ -431,51 +432,144 @@ var GlobalRoom = (function () {
 				return;
 			}
 			newSearch.rating = mmr;
-			self.addSearch(newSearch, user);
+			self.addSearch(newSearch, user, formatid);
 		});
 	};
-	GlobalRoom.prototype.matchmakingOK = function (search1, search2, user1, user2) {
+	GlobalRoom.prototype.matchmakingOK = function (search1, search2, user1, user2, formatid) {
 		// users must be different
-		if (user1 === user2) return false;
+		if (user1 === user2) return null;
 
 		// users must have different IPs
-		if (user1.latestIp === user2.latestIp) return false;
+		if (user1.latestIp === user2.latestIp) return null;
 
-		// users must not have been matched immediately previously
-		if (user1.lastMatch === user2.userid || user2.lastMatch === user1.userid) return false;
+		// users must not have been matched recently
+		if (user1.lastSearchFoes.indexOf(user2.userid) >= 0 || user2.lastSearchFoes.indexOf(user1.userid) >= 0) return null;
 
 		// search must be within range
-		var searchRange = 100, formatid = search1.formatid, elapsed = Math.abs(search1.time - search2.time);
+		var searchRange = 100, elapsed = Math.abs(search1.time - search2.time);
 		if (formatid === 'ou' || formatid === 'oucurrent' || formatid === 'randombattle') searchRange = 50;
 		searchRange += elapsed / 300; // +1 every .3 seconds
 		if (searchRange > 300) searchRange = 300;
 		if (Math.abs(search1.rating - search2.rating) > searchRange) return false;
 
-		user1.lastMatch = user2.userid;
-		user2.lastMatch = user1.userid;
 		return true;
 	};
-	GlobalRoom.prototype.addSearch = function (newSearch, user) {
-		if (!user.connected) return;
-		for (var i = 0; i < this.searchers.length; i++) {
-			var search = this.searchers[i];
-			var searchUser = Users.getExact(search.userid);
-			if (!searchUser || !searchUser.connected) {
-				this.searchers.splice(i, 1);
-				i--;
-				continue;
+	GlobalRoom.prototype.addSearch = function (newSearch, user, formatid) {
+		// Filter racing conditions
+		if (!user.connected || user !== Users.getExact(user.userid)) return;
+		if (user.searching[formatid]) return;
+
+		if (!this.searches[formatid]) this.searches[formatid] = [];
+		var formatSearches = this.searches[formatid];
+		var totalSearches = formatSearches.length;
+		var rating = newSearch.rating;
+
+		// Figure out the placement using binary search
```

Comment:
That's if you only try to pair the two topmost players. But what if they have just been matched? Then you have to find an indeterminate amount of top players, and at that point we are just sorting the whole list.